README.md
# SeqVec

Repository for the paper [Modeling aspects of the language of life through transfer-learning protein sequences](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-3220-8).
Holds pre-trained SeqVec model for creating embeddings for amino acid sequences. Also, contains checkpoint for fine-tuning.

# Abstract
**Background**: One common task in Computational Biology is the prediction of aspects of protein function and structure from their amino acid sequence. For 26 years, most state-of-the-art approaches toward this end have been marrying machine learning and evolutionary information. The retrieval of related proteins from ever growing sequence databases is becoming so time-consuming that the analysis of entire proteomes becomes challenging. On top, evolutionary information is less powerful for small families, e.g. for proteins from the Dark Proteome.

**Results**: We introduce a novel way to represent protein sequences as continuous vectors (embeddings) by using the deep bi-directional model ELMo taken from natural language processing (NLP). The model has effectively captured the biophysical properties of protein sequences from unlabeled big data (UniRef50). After training, this knowledge is transferred to single protein sequences by predicting relevant sequence features. We refer to these new embeddings as SeqVec (Sequence-to-Vector) and demonstrate their effectiveness by training simple convolutional neural networks on existing data sets for two completely different prediction tasks. At the per-residue level, we significantly improved secondary structure (for NetSurfP-2.0 data set: Q3=79%±1, Q8=68%±1) and disorder predictions (MCC=0.59±0.03) over methods not using evolutionary information. At the per-protein level, we predicted subcellular localization in ten classes (for DeepLoc data set: Q10=68%±1) and distinguished membrane- bound from water-soluble proteins (Q2= 87%±1). All results built upon the embeddings gained from the new tool SeqVec neither explicitly nor implicitly using evolutionary information. Nevertheless, it improved over some methods using such information. Where the lightning-fast HHblits needed on average about two minutes to generate the evolutionary information for a target protein, SeqVec created the vector representation on average in 0.03 seconds.

**Conclusion**: We have shown that transfer learning can be used to capture biochemical or biophysical properties of protein sequences from large unlabeled sequence databases. The effectiveness of the proposed approach was showcased for different prediction tasks using only single protein sequences. SeqVec embeddings enable predictions that outperform even some methods using evolutionary information. Thus, they prove to condense the underlying principles of protein sequences. This might be the first step towards competitive predictions based only on single protein sequences.

# t-SNE projections of SeqVec
![2D t-SNE projections](seqvec_tsne.png "2D t-SNE projections of SeqVec")
*2D t-SNE projections of unsupervised SeqVec embeddings highlight different realities of proteins and their constituent parts, amino acids.* Panels (b) to (d) are based on the same data set (Structural Classification of Proteins – extended (SCOPe) 2.07, redundancy reduced at 40%). For these plots, only subsets of SCOPe containing proteins with the annotation of interest (enzymatic activity (c) and kingdom (d)) may be displayed. **Panel (a)**: the embedding space confirms: the 20 standard amino acids are clustered according to their biochemical and biophysical properties, i.e. hydrophobicity, charge or size. The unique role of Cysteine (C, mostly hydrophobic and polar) is conserved. **Panel (b)**: SeqVec embeddings capture structural information as annotated in the main classes in SCOPe without ever having been explicitly trained on structural features. **Panel (c)**: many small, local clusters share function as given by the main classes in the Enzyme Commission Number (E.C.). **Panel (d)**: similarly, small, local clusters represent different kingdoms of life.

# Model availability
The ELMo model trained on UniRef50 (=SeqVec) is available at:
[SeqVec-model](https://rostlab.org/~deepppi/seqvec.zip)

The checkpoint for the pre-trained model is available at:
[SeqVec-checkpoint](https://rostlab.org/~deepppi/seqvec_checkpoint.tar.gz)

# Installation

```
pip install seqvec
```

We are working on a python package with more embedders and a commong interface to them; see [bio_embeddings](https://github.com/sacdallago/bio_embeddings)

# Example

In the [bio_embeddings](https://github.com/sacdallago/bio_embeddings) github repo, you can find examples in the `notebooks` folder.

For a general example on how to extract embeddings using ELMo, please check the
official ELMo implementation: [ELMo-Tutorial](https://github.com/allenai/bilm-tf)

You can compute embeddings for a fasta file with the `seqvec` command. Add `--protein True` to get an embedding per protein instead of per residue.

```
seqvec -i sequences.fasta -o embeddings.npz
```

Load the embeddings with numpy:

```python
import numpy as np
data = np.load("embeddings.npz")  # type: Dict[str, np.ndarray]
```

If you specify `.npy` as output format (e.g. with `-o embeddings.npy`), the script will save the embeddings as an numpy array and the corresponding identifiers (as extracted from the header line in the fasta file) in a json file besides it. The sorting in the json file corresponds to the indexing in the npy file. The npy file can be loaded via:

```python
import json
import numpy as np

data = np.load("embeddings.npy") # shape=(n_proteins,)
with open("embeddings.json") as fp:
    labels = json.load(fp)
```

**How to integrate the embedder into an existing workflow:**


Load pre-trained model:

```python
from allennlp.commands.elmo import ElmoEmbedder
from pathlib import Path

model_dir = Path('path/to/pretrained/SeqVec_directory')
weights = model_dir / 'weights.hdf5'
options = model_dir / 'options.json'
embedder = ElmoEmbedder(options,weights, cuda_device=0) # cuda_device=-1 for CPU
```

Get embedding for amino acid sequence:

```python
seq = 'SEQWENCE' # your amino acid sequence
embedding = embedder.embed_sentence(list(seq)) # List-of-Lists with shape [3,L,1024]
```

Batch embed sequences:

```python
seq1 = 'SEQWENCE' # your amino acid sequence
seq2 = 'PROTEIN'
seqs = [list(seq1), list(seq2)]
seqs.sort(key=len) # sorting is crucial for speed
embedding = embedder.embed_sentences(seqs) # returns: List-of-Lists with shape [3,L,1024]
```

Get 1024-dimensional embedding for per-residue predictions:

```python
import torch
residue_embd = torch.tensor(embedding).sum(dim=0) # Tensor with shape [L,1024]
```

Get 1024-dimensional embedding for per-protein predictions:
```python
protein_embd = torch.tensor(embedding).sum(dim=0).mean(dim=0) # Vector with shape [1024]
```
# FAQ
**Torch version conflict**
If you encounter a version conflict while pip-installing seqvec (```ERROR: No matching distribution found for torch<1.3,>=1.2 (from seqvec)```), creating a new conda-environment with Python 3.7 can resolve your issue.

**Slow embedding of very long sequences**
We've added an option which automatically falls back to CPU mode if even single-sequence processing fails on GPU due to memory problems. While this allows you to embed also very long proteins, e.g. Q8WZ42 (Titin, ~35k residues), it slows down the embedding process.


# Web-service for Predictions based on SeqVec
[SeqVec predictions - Chris' Protein properties](https://embed.protein.properties/)

# Bibtex-Reference
```
@article{heinzinger2019modeling,
  title={Modeling aspects of the language of life through transfer-learning protein sequences},
  author={Heinzinger, Michael and Elnaggar, Ahmed and Wang, Yu and Dallago, Christian and Nechaev, Dmitrii and Matthes, Florian and Rost, Burkhard},
  journal={BMC bioinformatics},
  volume={20},
  number={1},
  pages={723},
  year={2019},
  publisher={Springer}
}
```


Changelog.md
# Changelog

## 0.4.0 - 2020-07-23

 * Instead of `--protein 1` it's now simply `--protein`
 * Automatic CPU fallback for sequences to long for the GPU

## 0.3.0 - 2020-05-14

 * Changed `--no-sum-layers` to `--layer` with options `sum`, `all`, `CNN`, `LSTM1` and `LSTM2`. The default remains summing up all three layer.

tests/test_fallbacks.py
from pathlib import Path
from typing import List, Tuple
from unittest import mock

import numpy

from seqvec.seqvec import get_embeddings

# lengths, cpu, success
LogType = List[Tuple[List[int], bool, bool]]


class MockElmoMemory:
    cpu: bool
    log: LogType

    def __init__(self, cpu: bool, log: LogType):
        self.cpu = cpu
        self.log = log

    def embed_batch(self, batch: List[List[str]]) -> List[numpy.ndarray]:
        lengths = [len(sequence) for sequence in batch]
        if not self.cpu:
            if sum(len(i) for i in batch) > 18:
                self.log.append((lengths, False, False))
                raise RuntimeError(f"Too big for the GPU: {[len(i) for i in batch]}")
            else:
                self.log.append((lengths, False, True))
                return [numpy.zeros((3, length, 1024)) for length in lengths]
        else:
            assert (
                sum(len(i) for i in batch) > 18
            ), "This could have been done on the GPU"
            self.log.append((lengths, True, True))
            return [numpy.zeros((3, length, 1024)) for length in lengths]

    def embed_sentence(self, sentence: List[str]) -> numpy.ndarray:
        # That's what the original elmo does
        return self.embed_batch([sentence])[0]


def test_fallbacks(caplog):
    """ Check that the fallbacks to single sequence processing and/or the CPU are working.

    batchsize is 15, actual GPU limit 18, i.e. real values divided by 1000

    Procedure:
     * [7, 7, 7] Fails, passes with single sequence processing
     * [7, 8] Passes
     * [20] Fails, fails with single sequence processing, passes on the CPU
    """
    elmo_log: LogType = []
    with mock.patch(
        "seqvec.seqvec.get_elmo_model", lambda _, cpu: MockElmoMemory(cpu, elmo_log)
    ):
        fasta_file = Path("test-data/fallback_test_sequences.fasta")
        model_dir = Path()
        embeddings_generator = get_embeddings(
            fasta_file, model_dir, batchsize=15, layer="all", id_field=0
        )
        list(embeddings_generator)

    assert elmo_log == [
        ([7, 7, 7], False, False),
        ([7], False, True),
        ([7], False, True),
        ([7], False, True),
        ([7, 8], False, True),
        ([20], False, False),
        ([20], False, False),
        ([20], True, True),
    ]

    assert caplog.messages == [
        "Error processing batch of 3 sequences: Too big for the GPU: [7, 7, 7]",
        "Sequences in the failing batch: ['M7', 'I7', 'L7']",
        "Starting single sequence processing",
        "Error processing batch of 1 sequences: Too big for the GPU: [20]",
        "Sequences in the failing batch: ['T20']",
        "Starting single sequence processing",
        "RuntimeError for T20 with 20 residues: Too big for the GPU: [20]",
        "Single sequence processing failed. Switching to CPU now. This slows down the embedding process.",
    ]


tests/test_embedder_package.py
import json
from pathlib import Path
from tempfile import TemporaryDirectory
from typing import List
from unittest import mock

import numpy
import numpy as np
import pytest
from numpy import ndarray

from seqvec.seqvec import get_embeddings
from seqvec.seqvec import save_from_generator

SPLIT_CHAR = "|"
ID_FIELD = 1
CPU_FLAG = False
SUM_LAYERS = "sum"
BATCHSIZE = 15000
PER_PROTEIN = True


class MockElmoEmbedder:
    embedded_batch: List[ndarray]
    embedded_sentence: ndarray

    def __init__(self):
        self.embedded_batch = [
            numpy.load("test-data/embedded_batch0.npy"),
            numpy.load("test-data/embedded_batch1.npy"),
        ]
        self.embedded_sentence = numpy.load("test-data/embedded_sentence.npy")

    def embed_batch(self, batch: List[List[str]]) -> List[numpy.ndarray]:
        if len(batch) == 2:  # First call
            return self.embedded_batch
        else:  # Second call
            raise RuntimeError("Go into single-sequence processing due to OOM")

    def embed_sentence(self, _sentence: List[str]) -> numpy.ndarray:
        return self.embedded_sentence


def get_elmo_model_mock(_model_dir: Path, _cpu: bool) -> MockElmoEmbedder:
    return MockElmoEmbedder()


@pytest.mark.skip
def test_embedder():
    with TemporaryDirectory() as temp_dir:
        seq_dir = Path("test-data/sequences.fasta")
        path = Path(temp_dir) / "embeddings.npy"
        model_dir = Path("test-cache")
        embeddings_generator = get_embeddings(
            seq_dir,
            model_dir,
            SPLIT_CHAR,
            ID_FIELD,
            CPU_FLAG,
            SUM_LAYERS,
            BATCHSIZE,
            PER_PROTEIN,
        )
        save_from_generator(path, PER_PROTEIN, embeddings_generator)
        expected = np.load("test-data/embeddings.npy")
        actual = np.load(Path(temp_dir) / "embeddings.npy")
        assert np.allclose(expected, actual)

        expected = json.loads(Path("test-data/embeddings.json").read_text())
        actual = json.loads(Path(temp_dir).joinpath("embeddings.json").read_text())
        assert expected == actual


def test_single_sequence_processing():
    """ Check that the single sequence processing also performs the resizing postprocessing """
    with mock.patch("seqvec.seqvec.get_elmo_model", get_elmo_model_mock):
        seq_dir = Path("test-data/single_sequence_processing.fasta")
        model_dir = Path("test-cache")
        embeddings_generator = get_embeddings(
            seq_dir, model_dir, cpu=True, batchsize=400, layer="all"
        )
        for key, value in list(embeddings_generator):
            assert value.shape[1:] == (3072,)


seqvec/seqvec.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse
import gzip
import json
import logging
import urllib.request
from pathlib import Path
from typing import Dict, List, Tuple, Generator

import h5py
import numpy as np
import torch
from allennlp.commands.elmo import ElmoEmbedder
from tqdm import tqdm

logger = logging.getLogger(__name__)

EmbedderReturnType = Generator[Tuple[str, np.ndarray], None, None]


def get_elmo_model(model_dir: Path, cpu: bool) -> ElmoEmbedder:
    weights_path = model_dir / "weights.hdf5"
    options_path = model_dir / "options.json"

    # if no pre-trained model is available, yet --> download it
    if not (weights_path.exists() and options_path.exists()):
        logger.info(
            "No existing model found. Start downloading pre-trained SeqVec (~360MB)..."
        )

        Path.mkdir(model_dir, exist_ok=True)
        repo_link = "http://rostlab.org/~deepppi/embedding_repo/embedding_models/seqvec"
        options_link = repo_link + "/options.json"
        weights_link = repo_link + "/weights.hdf5"
        urllib.request.urlretrieve(options_link, str(options_path))
        urllib.request.urlretrieve(weights_link, str(weights_path))

    cuda_device = 0 if torch.cuda.is_available() and not cpu else -1
    logger.info("Loading the model")
    # The string casting comes from a typing bug in allennlp
    # https://github.com/allenai/allennlp/pull/3358
    return ElmoEmbedder(
        weight_file=str(weights_path),
        options_file=str(options_path),
        cuda_device=cuda_device,
    )


def read_fasta(
    sequences: Dict[str, str], fasta_path: Path, split_char: str, id_field: int
):
    """ Reads in fasta file containing multiple sequences.
    Adds all sequences to the `sequences` dictionary.
    """

    if fasta_path.suffix == ".gz":
        handle = gzip.open(str(fasta_path), "rt")
    else:
        handle = fasta_path.open()

    with handle as fasta_f:
        for line in fasta_f:
            # get uniprot ID from header and create new entry
            if line.startswith(">"):
                if id_field == -1:
                    uniprot_id = line.replace(">", "").strip()
                else:
                    uniprot_id = (
                        line.replace(">", "").strip().split(split_char)[id_field]
                    )
                sequences[uniprot_id] = ""
            else:
                # repl. all whie-space chars and join seqs spanning multiple lines
                sequences[uniprot_id] += "".join(line.split()).upper()


def read_fasta_file(seq_dir: Path, split_char: str, id_field: int) -> Dict[str, str]:
    seq_dict = dict()
    """ Read in FASTA file """
    if seq_dir.is_file():  # if single fasta file should be processed
        read_fasta(seq_dict, seq_dir, split_char, id_field)
    else:  # if a directory was provided: read all files
        assert seq_dir.is_dir(), f"'{seq_dir}' is neither a file nor a directory"
        for seq_path in seq_dir.glob("**/*fasta*"):
            read_fasta(seq_dict, seq_path, split_char, id_field)
    return seq_dict


def process_embedding(
    embedding: np.ndarray, per_protein: bool, layer: str
) -> np.ndarray:
    """
    Direct output of ELMo has shape (3,L,1024), with L being the protein's
    length, 3 being the number of layers used to train SeqVec (1 CharCNN, 2 LSTMs)
    and 1024 being a hyperparameter chosen to describe each amino acid.
    When a representation on residue level is required, you can sum
    over the first dimension, resulting in a tensor of size (L,1024),
    or just extract a specific layer.
    If you want to reduce each protein to a fixed-size vector, regardless of its
    length, you can average over dimension L.
    """
    if layer == "sum":
        # sum over residue-embeddings of all layers (3k->1k)
        embedding = embedding.sum(axis=0)
    elif layer == "CNN":
        embedding = embedding[0]
    elif layer == "LSTM1":
        embedding = embedding[1]
    elif layer == "LSTM2":
        embedding = embedding[2]
    else:
        # Stack the layer (3,L,1024) -> (L,3072)
        embedding = np.concatenate(embedding, axis=1)
    if per_protein:  # if embeddings are required on the level of whole proteins
        embedding = embedding.mean(axis=0)
    return embedding


_cpu_elmo_model = None  # Cache the CPU model across invocations


def embed_with_fallback(
    batch: List[Tuple[str, str]], model: ElmoEmbedder, model_dir: Path,
) -> EmbedderReturnType:
    """ Tries to get the embeddings in this order:
      * Full batch GPU
      * Single Sequence GPU
      * Single Sequence CPU

    Single sequence processing is done in case of runtime error due to
    a) very long sequence or b) too large batch size
    If this fails, you might want to consider lowering batchsize and/or
    cutting very long sequences into smaller chunks

    Returns unprocessed embeddings
    """
    global _cpu_elmo_model

    # create List[List[str]] for batch-processing of ELMo
    tokens = [list(seq) for _, seq in batch]
    batch_ids = [identifier for identifier, _ in batch]

    try:  # try to get the embedding for the current sequence
        with torch.no_grad():
            embeddings = model.embed_batch(tokens)
        assert len(batch) == len(embeddings)
        for sequence_id, embedding in zip(batch_ids, embeddings):
            yield sequence_id, embedding
    except RuntimeError as e:
        logger.error("Error processing batch of {} sequences: {}".format(len(batch), e))
        logger.error("Sequences in the failing batch: {}".format(batch_ids))
        logger.error("Starting single sequence processing")

        for sample_id, seq in batch:
            try:
                with torch.no_grad():
                    embedding = model.embed_sentence(list(seq))
                yield sample_id, embedding
            except RuntimeError as e:
                logger.error(
                    "RuntimeError for {} with {} residues: {}".format(
                        sample_id, len(seq), e
                    )
                )
                logger.error(
                    "Single sequence processing failed. Switching to CPU now. "
                    + "This slows down the embedding process."
                )
                if not _cpu_elmo_model:
                    _cpu_elmo_model = get_elmo_model(model_dir, cpu=True)
                with torch.no_grad():
                    embedding = _cpu_elmo_model.embed_sentence(list(seq))
                yield sample_id, embedding


def get_embeddings(
    seq_dir: Path,
    model_dir: Path,
    split_char: str = "|",
    id_field: int = 1,
    cpu: bool = False,
    layer: str = "sum",
    batchsize: int = 15000,
    per_protein: bool = False,
) -> EmbedderReturnType:
    """ Lazily generate all embeddings.

    You can use this function if you want to do postprocessing or need a custom output format.
    """
    seq_dict = read_fasta_file(seq_dir, split_char, id_field)

    # Sort sequences
    # Sorting sequences according to length is crucial for speed as batches
    # of proteins with similar size increase throughput.
    seq_dict = sorted(seq_dict.items(), key=lambda kv: len(seq_dict[kv[0]]))

    logger.info("Total number of sequences: {}".format(len(seq_dict)))

    model = get_elmo_model(model_dir, cpu)

    batch = list()
    length_counter = 0

    for index, (identifier, sequence) in enumerate(
        tqdm(seq_dict)
    ):  # for all sequences in the set

        # append sequence to batch and sum amino acids over proteins in batch
        batch.append((identifier, sequence))
        length_counter += len(sequence)

        # Transform list of batches to embeddings
        # if a) max. number of chars. for a batch is reached,
        # if b) sequence is longer than half batchsize (avoids RuntimeError for very long seqs.)
        # if c) the last sequence is reached
        if not (
            length_counter > batchsize
            or len(sequence) > batchsize / 2
            or index == len(seq_dict) - 1
        ):
            continue

        # Actually compute embeddings and postprocess
        for sequence_id, embedding in embed_with_fallback(batch, model, model_dir):
            yield sequence_id, process_embedding(embedding, per_protein, layer)

        # Reset batch
        batch = list()
        length_counter = 0


def save_from_generator(
    emb_path: Path,
    per_protein: bool,
    the_generator: Generator[Tuple[str, np.ndarray], None, None],
):
    if emb_path.suffix == ".h5":
        with h5py.File(str(emb_path), "w") as hf:
            for sequence_id, embedding in the_generator:
                if emb_path.suffix == ".h5":
                    # noinspection PyUnboundLocalVariable
                    hf.create_dataset(sequence_id, data=embedding)
    elif emb_path.suffix == ".npz" or emb_path.suffix == ".npy":
        if emb_path.suffix == ".npy" and not per_protein:
            raise RuntimeError(
                "You need to sum up per protein (`--protein True`) to save as .npy array"
            )

        emb_dict = dict()
        for sequence_id, embedding in the_generator:
            if embedding is None:
                # The generator code already showed an error
                continue
            emb_dict[sequence_id] = embedding

        if not emb_dict:
            raise RuntimeError("Embedding dictionary is empty!")
        logger.info("Total number of embeddings: {}".format(len(emb_dict)))

        if emb_path.suffix == ".npy":
            label_file = emb_path.with_suffix(".json")
            logger.info(f"Writing embeddings to {emb_path} and the ids to {label_file}")
            # save elmo representations
            with label_file.open("w") as id_file:
                json.dump(list(emb_dict.keys()), id_file)
            # noinspection PyTypeChecker
            np.save(emb_path, np.asarray(list(emb_dict.values())))
        else:
            logger.info(f"Writing embeddings to {emb_path}")
            # With checked that the suffix can only be .npz
            np.savez(emb_path, **emb_dict)
    else:
        raise RuntimeError(
            f"The output file must end with .npz, .npy or .h5,"
            f"but the path you provided ends with '{emb_path.suffix}'"
        )


def create_arg_parser():
    """ Creates and returns the ArgumentParser object. """

    # Instantiate the parser
    parser = argparse.ArgumentParser(
        description=(
            "seqvec.py creates ELMo embeddings for a given text "
            + " file containing sequence(s) in FASTA-format."
        )
    )

    # Path to fasta file (required)
    # noinspection PyTypeChecker
    parser.add_argument(
        "-i",
        "--input",
        required=True,
        type=Path,
        help="A path to a fasta-formatted text file containing protein sequence(s)."
        + "Can also be a directory holding multiple fasta files.",
    )

    # Path for writing embeddings (required)
    # noinspection PyTypeChecker
    parser.add_argument(
        "-o",
        "--output",
        required=True,
        type=Path,
        help="A path to a file for saving the created embeddings. "
        + "By default, a HDF (.h5) file will be written which should also be indicated by the chosen filename."
        + "Only if you create per-protein embeddings, you can also write to numpy formats, i.e. .npy or .npz, which again should be indicated by the chosen filename."
        + "If you choose to write a .npy file, a .json file with the sequence ids will be created next to the .npy file.",
    )

    # Path to model (optional)
    # noinspection PyTypeChecker
    parser.add_argument(
        "--model",
        type=Path,
        default=Path.cwd() / "model",
        help="A path to a directory holding a pre-trained ELMo model. "
        + "If the model is not found in this path, it will be downloaded automatically."
        + "The file containing the weights of the model must be named weights.hdf5."
        + "The file containing the options of the model must be named options.json",
    )

    # Create embeddings for a single protein or for all residues within a protein
    parser.add_argument(
        "--protein",
        action="store_true",
        default=False,
        help="Flag (no arguments needed) for summarizing embeddings from residue level to protein level "
        + "via averaging. Default: Not set (False)",
    )

    # Number of residues within one batch
    parser.add_argument(
        "--batchsize",
        type=int,
        default=15000,
        help="Number of residues which need to be accumulated before starting batch "
        + "processing. If you encounter an OutOfMemoryError, lower this value. Default: 15000",
    )

    # Character for splitting fasta header
    parser.add_argument(
        "--split-char",
        type=str,
        default="|",
        help="The character for splitting the FASTA header in order to retrieve "
        + "the protein identifier. Should be used in conjunction with --id. "
        + "Default: '|' ",
    )

    # Field index for protein identifier in fasta header after splitting with --split-char
    parser.add_argument(
        "--id",
        type=int,
        default=1,
        help="The zero based index for the uniprot identifier field after splitting the "
        + "FASTA header after each symbole in ['|', '#', ':', ' ']. "
        + "Use -1 to deactivate splitting. "
        + "Default: 1",
    )

    # Whether to use CPU or GPU
    parser.add_argument(
        "--cpu",
        action="store_true",
        default=False,
        help="Flag for using CPU to compute embeddings. Default: False",
    )
    parser.add_argument(
        "--layer",
        dest="layer",
        choices=["sum", "all", "CNN", "LSTM1", "LSTM2"],
        default="sum",
        help="Decide whether to `sum` up the layers (1024 dimensions), concatenate `all` of them (3072 dimensions) or "
        "select a specific layer (`CNN`, `LSTM1` or `LSTM2`). Defaults to `sum`",
    )

    parser.add_argument(
        "--silent",
        action="store_true",
        default=False,
        help="Embedder gives some information while processing. Default: True",
    )
    return parser


def main():
    parser = create_arg_parser()

    args = parser.parse_args()
    seq_dir = args.input
    emb_path = args.output
    model_dir = args.model
    split_char = args.split_char
    id_field = args.id
    cpu_flag = args.cpu
    per_prot = args.protein
    batchsize = args.batchsize
    verbose = not args.silent
    layer = args.layer

    if verbose:
        # Otherwise the default level is warning
        logging.basicConfig(level=logging.INFO, format="%(asctime)s %(message)s")

    device_name = "GPU" if torch.cuda.is_available() and not cpu_flag else "CPU"
    logger.info(f"Running on the {device_name}")

    embeddings_generator = get_embeddings(
        seq_dir, model_dir, split_char, id_field, cpu_flag, layer, batchsize, per_prot,
    )
    save_from_generator(emb_path, per_prot, embeddings_generator)


if __name__ == "__main__":
    main()
